### PR TITLE
feat(http): implement HTTP basic auth

### PR DESCRIPTION
**Pull request by jerrinot**
This PR allows configuring an HTTP server to require clients to authenticate via HTTP Basic.
It uses a single {username,password} pair hard-coded in a server configuration. 

This PR amends the default `server.conf` with this:
```
# Uncomment to enable authentication for HTTP server
#http.user=admin
#http.password=quest
```
Uncommenting these 2 lines enables authentication. There is no explicit switch to enable/disable configuration. The mere existence of these properties (with non-empty values) implicitly enables authentication. This is to keep property names consistent with PgWire configuration - it uses `pg.user` and  `pg.password`. 

When HTTP authentication is enabled, even the [Minimal HTTP server](https://questdb.io/docs/operations/logging-metrics/#minimal-http-server) used for healthchecks requires authentication. This is not always desirable; setting a property `http.health.check.authentication.required` to `false` allow anonymous access. 

It's worth noting that HTTP Basic without TLS should only be used on trusted networks.

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:disappointed: **fail** : 19 / 53 (35.85%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/http/StaticHttpAuthenticator.java|0|10|00.00%|
|:large_blue_circle:|io/questdb/cutlass/http/DefaultHttpServerConfiguration.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/cutlass/http/StaticHttpAuthenticatorFactory.java|0|19|00.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|4|5|80.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|4|5|80.00%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/RejectProcessorImpl.java|6|7|85.71%|
|:large_blue_circle:|io/questdb/cutlass/http/HttpConstants.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|2|2|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 50 / 59 (84.75%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/http/DefaultHttpServerConfiguration.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|7|11|63.64%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/RejectProcessorImpl.java|6|7|85.71%|
|:large_blue_circle:|io/questdb/cutlass/http/StaticHttpAuthenticatorFactory.java|17|19|89.47%|
|:large_blue_circle:|io/questdb/cutlass/http/HttpConstants.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/StaticHttpAuthenticator.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|2|2|100.00%|

File: core/src/main/resources/io/questdb/site/conf/server.conf [hunk 1]
```diff
@@ -30,6 +30,10 @@ config.validation.strict=true
 # IP address and port of HTTP server
 #http.net.bind.to=0.0.0.0:9000
 
+# Uncomment to enable authentication for HTTP server
```
Comment (puzpuzpuz) [inline core/src/main/resources/io/questdb/site/conf/server.conf:33]: nit: we should mention that the auth is the basic HTTP authentication

Comment (puzpuzpuz) [inline core/src/main/resources/io/questdb/site/conf/server.conf:33]: We should also add this to `pkg/ami/marketplace/assets/server.conf`

**Review by puzpuzpuz — APPROVED** [review]
LGTM. A couple of minor things may be improved.

**Comment by jerrinot** [comment]
@puzpuzpuz thanks for the review. I just pushed a new commit addressing your findings.

File: core/src/main/resources/io/questdb/site/conf/server.conf [hunk 2]
```diff
@@ -686,7 +690,7 @@ query.timeout.sec=60
 #pg.character.store.pool.capacity=64
 #pg.connection.pool.capacity=64
 #pg.password=quest
-#pg.user=admin
+#pg.user=adminBas
```
Comment (puzpuzpuz) [inline core/src/main/resources/io/questdb/site/conf/server.conf:693]: Do we need this change? The default seems to be `admin`.

Comment (jerrinot) [reply core/src/main/resources/io/questdb/site/conf/server.conf:693]: 🤦

Comment (jerrinot) [reply core/src/main/resources/io/questdb/site/conf/server.conf]: thank you! 🦅 👁️ !

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 50 / 59 (84.75%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/http/DefaultHttpServerConfiguration.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|7|11|63.64%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/RejectProcessorImpl.java|6|7|85.71%|
|:large_blue_circle:|io/questdb/cutlass/http/StaticHttpAuthenticatorFactory.java|17|19|89.47%|
|:large_blue_circle:|io/questdb/cutlass/http/HttpConstants.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/StaticHttpAuthenticator.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|2|2|100.00%|

**Review by puzpuzpuz — APPROVED** [review]

**Review by bluestreak01 — CHANGES_REQUESTED** [review]
we need to confirm Ent still works with this change

**Comment by glasstiger** [comment]
i think it should use the login screen instead of the basic auth realm, or support both (like enterprise).
without the login screen they have to play with the URL (http://username@host:9000) if they change the password, otherwise they cannot login.

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 49 / 57 (85.96%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/http/DefaultHttpServerConfiguration.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|7|11|63.64%|
|:large_blue_circle:|io/questdb/cutlass/http/StaticHttpAuthenticatorFactory.java|17|19|89.47%|
|:large_blue_circle:|io/questdb/cutlass/http/HttpConstants.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/StaticHttpAuthenticator.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/RejectProcessorImpl.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|2|2|100.00%|

**Comment by jerrinot** [comment]
@glasstiger thanks for your suggestion with the login page, that's a good one! 
I added https://github.com/questdb/ui/pull/300 so the login form is now displayed for QuestDB OSS with HTTP auth enabled.

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 49 / 57 (85.96%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/http/DefaultHttpServerConfiguration.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|7|11|63.64%|
|:large_blue_circle:|io/questdb/cutlass/http/StaticHttpAuthenticatorFactory.java|17|19|89.47%|
|:large_blue_circle:|io/questdb/cutlass/http/HttpConstants.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/StaticHttpAuthenticator.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/RejectProcessorImpl.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|2|2|100.00%|

**Review by glasstiger — APPROVED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 49 / 57 (85.96%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/http/DefaultHttpServerConfiguration.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|7|11|63.64%|
|:large_blue_circle:|io/questdb/cutlass/http/StaticHttpAuthenticatorFactory.java|17|19|89.47%|
|:large_blue_circle:|io/questdb/cutlass/http/HttpConstants.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/StaticHttpAuthenticator.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/RejectProcessorImpl.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|2|2|100.00%|

**Comment by glasstiger** [comment]
> we need to confirm Ent still works with this change

tested with ent, and getting an error when trying to login with non-admin user.
quite cryptic.
```
2024-06-20T15:33:55.512650Z E i.q.c.h.HttpConnectionContext internal error [fd=683, e=`
java.lang.ExceptionInInitializerError
	at com.questdb.security.AuthCrypto.<clinit>(AuthCrypto.java:80)
	at com.questdb.security.PasswordStoreImpl.verifyPassword(PasswordStoreImpl.java:240)
	at io.questdb.cutlass.pgwire.CombiningUsernamePasswordMatcher.verifyPassword(CombiningUsernamePasswordMatcher.java:39)
	at com.questdb.cutlass.http.EntHttpAuthenticator.authenticateBasic(EntHttpAuthenticator.java:127)
	at com.questdb.cutlass.http.EntHttpAuthenticator.doAuthenticate(EntHttpAuthenticator.java:181)
	at com.questdb.cutlass.http.EntHttpAuthenticator.authenticate(EntHttpAuthenticator.java:60)
	at io.questdb.cutlass.http.HttpConnectionContext.configureSecurityContext(HttpConnectionContext.java:482)
	at io.questdb.cutlass.http.HttpConnectionContext.handleClientRecv(HttpConnectionContext.java:868)
	at io.questdb.cutlass.http.HttpConnectionContext.handleClientOperation(HttpConnectionContext.java:256)
	at io.questdb.cutlass.http.HttpServer.handleClientOperation(HttpServer.java:307)
	at io.questdb.cutlass.http.HttpServer$1.lambda$$0(HttpServer.java:114)
	at io.questdb.network.AbstractIODispatcher.processIOQueue(AbstractIODispatcher.java:199)
	at io.questdb.cutlass.http.HttpServer$1.run(HttpServer.java:118)
	at io.questdb.mp.Worker.run(Worker.java:151)
Caused by: io.questdb.jar.jni.LoadException: Internal error: cannot find /com/questdb/bin/darwin-aarch64/libqdb_ent.dylib, broken package?
	at io.questdb.jar.jni.JarJniLoader.loadLib(JarJniLoader.java:54)
	at io.questdb.jar.jni.JarJniLoader.loadLib(JarJniLoader.java:101)
	at com.questdb.std.QdbEnt.<clinit>(QdbEnt.java:21)
	...14 more`]
```

**Review by bluestreak01 — APPROVED** [review]

**Pull request merged**